sympy/solvers/solveset.py

```diff
@@ -1827,22 +1881,28 @@ def _is_lambert(f, symbol):
 
     # total number of symbols in equation
     no_of_symbols = len([arg for arg in term_factors if arg.has(symbol)])
-    # total number of trigonometric terms in equation
-    no_of_trig = len([arg for arg in term_factors \
-        if arg.has(HyperbolicFunction, TrigonometricFunction)])
 
     if f.is_Add and no_of_symbols >= 2:
         # `log`, `HyperbolicFunction`, `TrigonometricFunction` should have symbols
         # and no_of_trig < no_of_symbols
         lambert_funcs = (log, HyperbolicFunction, TrigonometricFunction)
         if any(isinstance(arg, lambert_funcs)\
             for arg in term_factors if arg.has(symbol)):
-                if no_of_trig < no_of_symbols:
-                    return True
-        # here, `Pow`, `exp` exponent should have symbols
+            # total number of trig terms in equation
+            no_of_trig = len([arg for arg in term_factors \
+                if arg.has(HyperbolicFunction, TrigonometricFunction) and arg.has(symbol)])
```

Comment:
Why is this using `arg.has(HyperbolicFunction, TrigonometricFunction)` rather than
```
isinstance(arg, (HyperbolicFunction, TrigonometricFunction))
```

Reply:
I thought both have no difference .

Reply:
```python
In [91]: expr = sin(LambertW(x))

In [92]: expr
Out[92]: sin(W(x))

In [93]: expr.has(LambertW)
Out[93]: True

In [94]: isinstance(expr, LambertW)
Out[94]: False
```